scylla/src/transport/session_test.rs

```diff
@@ -33,6 +38,38 @@ use std::time::Duration;
 use tokio::net::TcpListener;
 use uuid::Uuid;
 
+#[cfg(not(feature = "cloud"))]
+type ConditionalSessionBuilderKind = DefaultMode;
+#[cfg(feature = "cloud")]
+type ConditionalSessionBuilderKind = CloudMode;
+
+/*
+  Creates a generic session builder based on conditional compilation configuration
+  For SessionBuilder of DefaultMode type, adds localhost to known hosts, as all of the tests
+  connect to localhost.
+*/
+fn create_new_session_builder() -> GenericSessionBuilder<ConditionalSessionBuilderKind> {
+    #[cfg(not(feature = "cloud"))]
```

Comment:
I'm not sure whether we should use features for this. It's a bit fragile to depend on some feature _not_ being enabled - I remember that there were some issue with one example requiring the "cloud" feature, and compiling tests enabled this feature for everybody, right?

Instead, I think that we should use conditional compilation. You can provide a `--cfg my-thing-here` parameter when compiling and add a bunch of attributes that enable/disable compilation based on those parameters - like you did with features as of now.

You could use a `--cfg cloud-tests` parameter. Then:

- Tests that are "cloud_agnostic" would use `create_new_session_builder()` which creates either a cloud or non-cloud session, depending on the parameter,
- Tests that require cloud would be compiled only if the parameter is set in commandline - so, they would be prepended with `#[cfg_attr(cloud_agnostic)]`,
- Tests that must not run with serverless would be decorated with `#[not(cfg_attr(cloud_agnostic))]`.

This method will work even if `cloud` feature is enabled in both cases, so I think it is safer.

Read more about conditional compilation here: https://doc.rust-lang.org/reference/conditional-compilation.html

Reply:
My initial approach was to use the `--cfg` flag, however, `cargo test` does not support it and using the `cloud` feature is meant to be a workaround to conditionally run tests. As the `--cfg` is a `rustc` flag, I can pass it through the `RUSTFLAGS` environment variable but this may affect the dependencies as well. So I found the workaround with a feature as a better approach.

Reply:
Passing `--cfg` via `RUSTFLAGS` is fine.

I don't think that any of the dependencies will react to `cloud_tests`, but if you are worried about that then we can prefix it with the name of the crate - so, `--cfg scylla_cloud_tests`. For example, tokio does exactly that - see here: https://github.com/tokio-rs/tokio/blob/master/.github/workflows/ci.yml